src/api/mod.rs

```diff
@@ -1,29 +1,160 @@
+        loop {
+            let res = lock.refresh_token(&api.api_root, &api.client).await;
+
+            match res {
+                Ok(_) => break,
+                Err(e) => {
+                    println!("{:?}", e);
+                }
+            }
+        }
+
+        drop(lock)
+    }
+}
+
+pub async fn refresh_token_worker(api: Arc<OtrApiBody>, receiver: Receiver<()>) {
+    tokio::select! {
+        _ = refresh_token_loop(api) => {}
+        _ = receiver => {}
+    }
+}
+
+pub struct OtrToken {
+    pub token: String,
+    pub refresh_token: String,
+    pub expire_in: u64
+}
+
+impl OtrToken {
+    /// Function that refresh token when called
+    pub async fn refresh_token(&mut self, api_root: &str, client: &Client) -> Result<(), Error> {
+        let link = format!("{}/v1/oauth/refresh?refreshToken={}", api_root, self.refresh_token);
+
+        let mut response: OAuthResponse = client
+            .post(link)
+            .header(CONTENT_TYPE, "application/json")
+            .send()
+            .await?
+            .json()
+            .await?;
+
+        response.token.insert_str(0, "Bearer ");
+
+        self.token = response.token;
+        self.refresh_token = response.refresh_token;
+        self.expire_in = response.expire_in;
+
+        Ok(())
+    }
+}
+
+pub struct OtrApiBody {
     client: Client,
-    token: String,
-    api_root: String
+    api_root: String,
+
+    /// Wrapped in RwLock because we need
+    /// a shared mutable access
+    token: RwLock<OtrToken>
+}
+
+pub struct OtrApiClient {
+    /// Wrapped in [`Arc`] because everything located in [`OtrApiBody`]
+    /// need to be accessed in diffrent threads (shared reference)
+    body: Arc<OtrApiBody>,
+
+    /// Channel that's get sended on Drop to shutdown refresh token worker
+    /// Why it's wrapped in `[Option]`? Because oneshot sender
+    /// consumes itself on send.
+    ///
+    /// Detailed explanation:
+    /// In Drop trait implementation we having a mutable reference
+    /// on our struct (where sender is located), so when `send()`
+    /// happens it consumes itself, but we still having that mutable
+    /// reference and because of this we getting a compile error
+    /// variabled getting moved.
+    /// Workaround is pretty simple:
+    /// Wrap [`Sender`] inside Option, so we can use [`std::mem::take`]
+    /// to replace our sender with default value (in our case in None)
+    /// and peacefully let sender consume itself :)
+    refresh_tx: Option<Sender<()>>
+}
+
+impl Drop for OtrApiClient {
+    fn drop(&mut self) {
+        if let Some(tx) = std::mem::take(&mut self.refresh_tx) {
+            // Dropping error because either `Ok` or `Err` indicates
+            // that worker and loop is stoped
+            // Ok - means channel is read and loop is stopped
+            // Err - means Receiver was somehow dropped beforehand
+            // which means that worker is not running under any
+            // circumstances
```

Comment:
```suggestion
            // Dropping error because either `Ok` or `Err` indicates
            // that the worker and loop are stopped.
            // Ok - means the channel is read and the loop is stopped.
            // Err - means the receiver was somehow dropped beforehand,
            // which means that the worker is not running under any circumstances.
```